File: android/app/src/main/java/com/macr120/mindhome/widgets/WidgetsPlugin.java
package com.macr120.mindhome.widgets;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.util.Base64;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import com.macr120.mindhome.R;

/**
 * Plugin Capacitor local «Widgets»: el puente entre el WebView (src/core/widgets/)
 * y los widgets del launcher. Java nunca decide negocio — guarda lo que la app
 * publica y le entrega lo que el widget encoló.
 */
@CapacitorPlugin(name = "Widgets")
public class WidgetsPlugin extends Plugin {

  /** Arranque en frío desde el toque en un widget: el Intent ya trae los extras. */
  @Override
  public void load() {
    guardarDestinoDe(getActivity().getIntent());
  }

  /** App ya viva (singleTask): el toque llega por onNewIntent. */
  @Override
  protected void handleOnNewIntent(Intent intent) {
    super.handleOnNewIntent(intent);
    guardarDestinoDe(intent);
  }

  private void guardarDestinoDe(Intent intent) {
    if (intent == null) return;
    String app = intent.getStringExtra("mph_app");
    String accion = intent.getStringExtra("mph_accion");
    if (app == null && accion == null) return;
    try {
      JSONObject destino = new JSONObject();
      if (app != null) destino.put("appId", app);
      if (accion != null) destino.put("accion", accion);
      String seccion = intent.getStringExtra("mph_seccion");
      if (seccion != null) destino.put("seccion", seccion);
      WidgetsStore.guardarDestino(getContext(), destino);
    } catch (JSONException ignorada) {
      return;
    }
    // Que una recreación de la Activity (rotación) no repita el destino.
    intent.removeExtra("mph_app");
    intent.removeExtra("mph_seccion");
    intent.removeExtra("mph_accion");
  }

  /** Destino del último toque en un widget ({} si no hay); leerlo lo consume. */
  @PluginMethod
  public void tomarDestinoPendiente(PluginCall call) {
    JSONObject destino = WidgetsStore.tomarDestino(getContext());
    JSObject res = new JSObject();
    if (destino != null) {
      if (destino.has("appId")) res.put("appId", destino.optString("appId"));
      if (destino.has("seccion")) res.put("seccion", destino.optString("seccion"));
      if (destino.has("accion")) res.put("accion", destino.optString("accion"));
    }
    call.resolve(res);
  }

  /** Guarda el snapshot, poda optimistas ya aplicados y repinta los widgets. */
  @PluginMethod
  public void publicarSnapshot(PluginCall call) {
    String json = call.getString("json");
    if (json == null) {
      call.reject("Falta json");
      return;
    }
    Context ctx = getContext();
    WidgetsStore.guardarSnapshot(ctx, json);
    WidgetsStore.podarOptimistas(ctx);
    repintarTodos(ctx);
    call.resolve();
  }

  /** Guarda la foto de la casa (JPEG en base64) y repinta su widget. */
  @PluginMethod
  public void publicarFotoCasa(PluginCall call) {
    String base64 = call.getString("base64");
    if (base64 == null) {
      call.reject("Falta base64");
      return;
    }
    Context ctx = getContext();
    try {
      WidgetsStore.guardarFotoCasa(ctx, Base64.decode(base64, Base64.DEFAULT));
    } catch (IOException | IllegalArgumentException e) {
      call.reject("No se pudo guardar la foto de la casa", e);
      return;
    }
    AppWidgetManager mgr = AppWidgetManager.getInstance(ctx);
    for (int id : ids(ctx, mgr, CasaWidgetProvider.class)) {
      CasaWidgetProvider.pintar(ctx, mgr, id);
    }
    call.resolve();
  }

  /** Qué widgets hay colocados: si ninguno, la app ni arma el snapshot. */
  @PluginMethod
  public void hayWidgets(PluginCall call) {
    Context ctx = getContext();
    AppWidgetManager mgr = AppWidgetManager.getInstance(ctx);
    boolean casa = ids(ctx, mgr, CasaWidgetProvider.class).length > 0;
    boolean hoy = ids(ctx, mgr, HoyWidgetProvider.class).length > 0;
    JSObject res = new JSObject();
    // El de chat no cuenta: es un lanzador puro, no consume datos de la app.
    res.put("hay", casa || hoy);
    res.put("casa", casa);
    call.resolve(res);
  }

  /** Devuelve la cola de acciones y la vacía (atómico en WidgetsStore). */
  @PluginMethod
  public void tomarAccionesPendientes(PluginCall call) {
    try {
      JSObject res = new JSObject();
      res.put("acciones", new JSArray(WidgetsStore.tomarAcciones(getContext())));
      call.resolve(res);
    } catch (JSONException e) {
      call.reject("Cola de acciones ilegible", e);
    }
  }

  private static int[] ids(Context ctx, AppWidgetManager mgr, Class<?> provider) {
    return mgr.getAppWidgetIds(new ComponentName(ctx, provider));
  }

  /** Repinta los widgets que viven del snapshot y recarga la lista de «Hoy». */
  static void repintarTodos(Context ctx) {
    AppWidgetManager mgr = AppWidgetManager.getInstance(ctx);
    for (int id : ids(ctx, mgr, CasaWidgetProvider.class)) {
      CasaWidgetProvider.pintar(ctx, mgr, id);
    }
    int[] hoy = ids(ctx, mgr, HoyWidgetProvider.class);
    for (int id : hoy) {
      HoyWidgetProvider.pintar(ctx, mgr, id);
    }
    mgr.notifyAppWidgetViewDataChanged(hoy, R.id.widget_hoy_lista);
  }
}

File: ios/App/App/WidgetsPlugin.swift
import Capacitor
import Foundation
import WidgetKit

/**
 * Plugin Capacitor local «Widgets»: el puente entre el WebView
 * (`src/core/widgets/`) y los widgets de la pantalla de inicio. Mismo contrato
 * que el de Android (`WidgetsPlugin.java`), método por método, para que
 * `src/core/widgets/plugin.ts` no tenga que saber en qué plataforma corre.
 *
 * Swift nunca decide negocio: guarda lo que la app publica y le entrega lo que
 * el widget encoló.
 */
@objc(WidgetsPlugin)
public class WidgetsPlugin: CAPPlugin, CAPBridgedPlugin {
  public let identifier = "WidgetsPlugin"
  public let jsName = "Widgets"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "publicarSnapshot", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "publicarFotoCasa", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "tomarAccionesPendientes", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "hayWidgets", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "tomarDestinoPendiente", returnType: CAPPluginReturnPromise),
  ]

  /// Los `kind` que declara cada widget en `MPHWidgets`.
  private static let kindCasa = "MPHCasa"
  private static let kindMisiones = "MPHMisiones"

  /// Guarda el snapshot, poda optimistas ya aplicados y repinta los widgets.
  @objc func publicarSnapshot(_ call: CAPPluginCall) {
    guard let json = call.getString("json") else {
      call.reject("Falta json")
      return
    }
    AlmacenWidgets.guardarSnapshot(json)
    AlmacenWidgets.podarOptimistas()
    WidgetCenter.shared.reloadAllTimelines()
    call.resolve()
  }

  /// Guarda la foto de la casa (JPEG en base64, sin cabecera `data:`).
  @objc func publicarFotoCasa(_ call: CAPPluginCall) {
    guard let base64 = call.getString("base64") else {
      call.reject("Falta base64")
      return
    }
    guard let datos = Data(base64Encoded: base64, options: .ignoreUnknownCharacters) else {
      call.reject("La foto de la casa no es base64 válido")
      return
    }
    do {
      try AlmacenWidgets.guardarFotoCasa(datos)
    } catch {
      call.reject("No se pudo guardar la foto de la casa", nil, error)
      return
    }
    WidgetCenter.shared.reloadTimelines(ofKind: Self.kindCasa)
    call.resolve()
  }

  /// Devuelve la cola de acciones y la vacía.
  @objc func tomarAccionesPendientes(_ call: CAPPluginCall) {
    let acciones = AlmacenWidgets.tomarAcciones().map { a -> [String: Any] in
      ["accionId": a.accionId, "id": a.id, "tipo": a.tipo,
       "fecha": a.fecha, "hecho": a.hecho, "ts": a.ts]
    }
    call.resolve(["acciones": acciones])
  }

  /**
   * Qué widgets hay colocados: si ninguno, la app ni arma el snapshot. El de
   * chat no cuenta —es un lanzador puro que no consume datos—, igual que en
   * Android.
   */
  @objc func hayWidgets(_ call: CAPPluginCall) {
    WidgetCenter.shared.getCurrentConfigurations { resultado in
      guard case let .success(infos) = resultado else {
        // Sin permiso o sin WidgetKit disponible: mejor no publicar que reventar.
        call.resolve(["hay": false, "casa": false])
        return
      }
      let casa = infos.contains { $0.kind == Self.kindCasa }
      let misiones = infos.contains { $0.kind == Self.kindMisiones }
      // Sin widget de casa colocado, el JPEG del contenedor no le sirve a nadie.
      // Android lo borra en `onDisabled`; WidgetKit no avisa de la retirada, así
      // que el momento equivalente es esta consulta, que la app ya hace sola.
      if !casa { AlmacenWidgets.borrarFotoCasa() }
      call.resolve(["hay": casa || misiones, "casa": casa])
    }
  }

  /// Destino del último toque en un widget ({} si no hay); leerlo lo consume.
  @objc func tomarDestinoPendiente(_ call: CAPPluginCall) {
    guard let destino = AlmacenWidgets.tomarDestino() else {
      call.resolve([:])
      return
    }
    var res: [String: Any] = [:]
    for clave in ["appId", "seccion", "accion"] where destino[clave] != nil {
      res[clave] = destino[clave]!
    }
    call.resolve(res)
  }
}
